File: frontend/src/types/order.ts
export type OrderItem = {
  id: string;
  product_id: string;
  vendor_id: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
};

export type Order = {
  id: string;
  status: "payment_pending" | "commande" | "preparation" | "livraison" | "recu" | "CLAIMED";
  payment_mode: "nita" | "amana";
  delivery_type: "standard" | "express_niamey";
  payment_reference: string | null;
  payment_status: "pending" | "paid";
  payment_confirmed_at: string | null;
  transaction_code: string | null;
  tracking_code: string | null;
  estimated_minutes: number;
  total_amount: number;
  currency: string;
  created_at: string;
  items: OrderItem[];
};

export type CheckoutPayload = {
  items: Array<{
    product_id: string;
    vendor_id: string;
    quantity: number;
    unit_price: number;
  }>;
  payment_mode: "nita" | "amana";
  delivery_type: "standard" | "express_niamey";
  transaction_code?: string;
  currency?: string;
};

export type ReceiptLink = {
  order_id: string;
  token: string;
  receipt_url: string;
  verify_url: string;
};

export type ReceiptItem = {
  product_id: string;
  product_name: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
};

export type Receipt = {
  order_id: string;
  customer_name: string;
  payment_mode: "nita" | "amana";
  payment_reference: string | null;
  payment_status: "pending" | "paid";
  currency: string;
  total_amount: number;
  transaction_code_masked: string | null;
  created_at: string;
  issued_at: string;
  items: ReceiptItem[];
  integrity_hash: string;
  verify_url: string;
  payment_url: string;
  platform_wallet_phone: string | null;
};

export type ReceiptVerifyPayload = {
  token: string;
  vendor_id?: string;
  gps?: string;
};

export type ReceiptVerifyResult = {
  order_id: string;
  status: "claimed" | "blocked";
  message: string;
  scanned_at: string | null;
};

export type PaymentIntent = {
  order_id: string;
  payment_mode: "nita" | "amana";
  payment_reference: string;
  amount: number;
  currency: string;
  payment_url: string;
  qr_payload: string;
  expires_in_seconds: number;
};

export type PaymentConfirmPayload = {
  provider_reference?: string;
  code_last4?: string;
};

export type PaymentConfirmResult = {
  order_id: string;
  payment_status: "pending" | "paid";
  order_status: Order["status"];
  message: string;
};

File: backend/app/schemas/order.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class CheckoutItemRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str = Field(min_length=1, max_length=36)
    vendor_id: str = Field(min_length=1, max_length=36)
    quantity: int = Field(ge=1, le=100)
    unit_price: float = Field(gt=0)


class CheckoutRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    items: list[CheckoutItemRequest] = Field(min_length=1)
    payment_mode: str = Field(default="nita", pattern="^(nita|amana)$")
    delivery_type: str = Field(default="standard", pattern="^(standard|express_niamey)$")
    transaction_code: str | None = Field(default=None, max_length=120)
    currency: str = Field(default="XOF", min_length=3, max_length=3)


class OrderItemResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    product_id: str
    vendor_id: str
    quantity: int
    unit_price: float
    subtotal: float


class OrderResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    status: str
    payment_mode: str
    delivery_type: str
    transaction_code: str | None
    tracking_code: str | None
    estimated_minutes: int
    total_amount: float
    currency: str
    created_at: datetime
    items: list[OrderItemResponse]


class ReceiptItemResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str
    product_name: str
    quantity: int
    unit_price: float
    subtotal: float


class ReceiptLinkResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_id: str
    token: str
    receipt_url: str
    verify_url: str


class ReceiptResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_id: str
    customer_name: str
    payment_mode: str
    currency: str
    total_amount: float
    transaction_code_masked: str | None
    created_at: datetime
    issued_at: datetime
    items: list[ReceiptItemResponse]
    integrity_hash: str
    verify_url: str


class ReceiptVerifyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    token: str = Field(min_length=32, max_length=2048)
    vendor_id: str | None = Field(default=None, max_length=36)
    gps: str | None = Field(default=None, max_length=120)


class ReceiptVerifyResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_id: str
    status: str
    message: str
    scanned_at: datetime | None
